### PR TITLE
Implement mock cosmos chain using `ibc-rs`, `tendermint-rs` and `basecoin` libs

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: informalsystems/hermes-sdk#24 

## Context

Given improvements made to `basecoin-rs` (turning it [into a library](https://github.com/informalsystems/basecoin-rs/pull/103) and moving the store types and implementations [into a separate](https://github.com/informalsystems/basecoin-rs/pull/127) crate from the application part), It became possible to implement a mock cosmos chain.

- Concerning the **consensus engine**, a simple block production engine has been used. Without needing to go through the entire voting, validation, transaction, etc. processes that a CometBFT engine requires. This is achieved using the `tendermint-rs` crates (specifically `tendermint-testgen`).

- As for the **application** itself, the necessary modules (Authentication, Bank, and IBC) have been imported from the `basecoin_app` crate. This has resulted in a **fully operational ibc-rs** that can be run in a lightweight in-memory manner.

- In terms of the **store** part, the benefit of the in-memory AVL store implementation within `basecoin_store` can now be realized, which enables proof verification.

- Within this PR, tests for creating and updating clients are included, and further tests can be added perhaps in separate PRs.

This implementation can serve as a foundation for developing a [new mock module in ibc-rs](https://github.com/cosmos/ibc-rs/issues/677). Additionally, it can be leveraged as an exemplary implementation for the [ongoing development of the ibc-rs](https://github.com/cosmos/ibc-rs/issues/743) doc portal.
_____

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [X] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer-cosmos-mock/src/contexts/chain.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+use std::sync::Arc;
+use std::sync::Mutex;
+
+use super::runtime::MockClock;
+use crate::contexts::runtime::MockRuntimeContext;
+use crate::traits::endpoint::BasecoinEndpoint;
+use crate::types::error::Error;
+use crate::types::status::ChainStatus;
+use crate::util::mutex::MutexUtil;
+
```
Comment (seanchen1991) [inline crates/relayer-cosmos-mock/src/contexts/chain.rs:19]:
```suggestion
/// Holds the necessary fields for querying a mock Cosmos
/// chain endpoint.
```

File: crates/relayer-cosmos-mock/src/impls/basecoin.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    async fn commit(&self) {
+        let mut modules = self.app.modules.write_access();
+
+        for IdentifiedModule { id, module } in modules.iter_mut() {
+            module
+                .store_mut()
+                .commit()
+                .expect("failed to commit to state");
+
+            let mut state = self.app.store.write_access();
```
Comment (seanchen1991) [inline crates/relayer-cosmos-mock/src/impls/basecoin.rs:88]: Small nit: could we not move `let mut state = self.app.store.write_access();` to the top of the function outside of the for loop and then not have to declare `let mut state` again in the rest of the function?

File: crates/relayer-cosmos-mock/src/impls/basecoin.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+        let mut modules = self.app.modules.write_access();
+
+        for IdentifiedModule { id: _, module } in modules.iter_mut() {
+            let event = module.begin_block(&last_block.signed_header.header);
+            events.extend(event);
+        }
+    }
+
+    /// Commits the chain state.
```
Comment (seanchen1991) [inline crates/relayer-cosmos-mock/src/impls/basecoin.rs:78]:
```suggestion
    /// Commits the chain state to the store.
```

File: crates/relayer-cosmos-mock/src/types/error.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        if let Some(expected) = &self.expected {
+            write!(f, ", Expected: {}", expected)?;
+        }
+
+        Ok(())
+    }
+}
+
```
Comment (seanchen1991) [inline crates/relayer-cosmos-mock/src/types/error.rs:108]:
```suggestion
/// Defines all the possible error types and their corresponding codes for 
/// interchain account errors.
```

File: crates/relayer-cosmos-mock/src/impls/basecoin.rs (hunk 2 above)
Comment (Farhad-Shabani) [reply crates/relayer-cosmos-mock/src/impls/basecoin.rs]: [cd438ec](https://github.com/informalsystems/hermes/pull/3566/commits/cd438ec5775774bf71457a103726b39bba8f7edc)

File: crates/relayer-cosmos-mock/src/contexts/basecoin.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    pub fn grow_blocks(&self) {
+        let mut blocks = self.blocks.acquire_mutex();
+
+        let validators = [
+            Validator::new("1").voting_power(40),
+            Validator::new("2").voting_power(30),
+            Validator::new("3").voting_power(30),
+        ];
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply crates/relayer-cosmos-mock/src/contexts/basecoin.rs]: Defined a new field for that so each chain can be set differently now.

File: crates/relayer-cosmos-mock/src/contexts/chain.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    pub fn get_current_timestamp(&self) -> Timestamp {
+        self.current_status.acquire_mutex().timestamp
+    }
+
+    pub fn get_current_height(&self) -> Height {
+        self.current_status.acquire_mutex().height
+    }
+
+    /// Keeps mock chain context tracking the latest status of the chain.
+    pub fn connect(&self) {
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply crates/relayer-cosmos-mock/src/contexts/chain.rs]: Right. Actually, was symptom of design. It is improved now by moving the entire chain status tracker under the `MockBasecoin` struct. So, don’t need this function anymore.

**Comment by soareschen** [comment]
This PR has been moved to https://github.com/informalsystems/hermes-sdk/pull/39

**Pull request closed without merging**